components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java

```diff
@@ -877,6 +878,166 @@ public User updateUser(User user, Map<String, Boolean> requiredAttributes) throw
+            Map<String, String> allSimpleMultiValuedClaims = new HashMap<>();
+
+            for (String simpleMultiValuedAttribute : allSimpleMultiValuedAttributes) {
+                allSimpleMultiValuedClaims.put(simpleMultiValuedAttribute, "");
+            }
+            Map<String, String> allSimpleMultiValuedClaimsList =
+                    SCIMCommonUtils.convertSCIMtoLocalDialect(allSimpleMultiValuedClaims);
+
+            // Check if username of the updating user existing in the userstore.
+            try {
+                String userStoreDomainFromSP = getUserStoreDomainFromSP();
+                SCIMResourceTypeSchema schema = SCIMResourceSchemaManager.getInstance().getUserResourceSchema();
+                User oldUser = this.getUser(user.getId(), ResourceManagerUtil.getAllAttributeURIs(schema));
+                if (userStoreDomainFromSP != null && !userStoreDomainFromSP
+                        .equalsIgnoreCase(IdentityUtil.extractDomainFromName(oldUser.getUserName()))) {
+                    String errorMessage =
+                            String.format("User :  %s is not belong to userstore %s. Hence user updating fail",
+                                    oldUser.getUserName(), userStoreDomainFromSP);
+                    throw new CharonException(errorMessage);
+                }
+                if (getUserStoreDomainFromSP() != null &&
+                        !UserCoreConstants.PRIMARY_DEFAULT_DOMAIN_NAME.equalsIgnoreCase(getUserStoreDomainFromSP())) {
+                    user.setUserName(IdentityUtil
+                            .addDomainToName(UserCoreUtil.removeDomainFromName(user.getUserName()),
+                                    getUserStoreDomainFromSP()));
+                }
+                // This is handled here as the IS is still not capable of updating the username via SCIM.
+                if (!StringUtils.equals(user.getUserName(), oldUser.getUserName())) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Failing the request as attempting to modify username. Old username: "
+                                + oldUser.getUserName() + ", new username: " + user.getUserName());
+                    }
+
+                    throw new BadRequestException("Attribute userName cannot be modified.",
+                            ResponseCodeConstants.MUTABILITY);
+                }
+            } catch (IdentityApplicationManagementException e) {
+                throw new CharonException("Error retrieving Userstore name. ", e);
+            }
+
+            boolean isExistingUser;
+            if (StringUtils.isNotEmpty(user.getId())) {
+                isExistingUser = carbonUM.isExistingUserWithID(user.getId());
+            } else {
+                isExistingUser = carbonUM.isExistingUser(user.getUserName());
+            }
+
+            if (!isExistingUser) {
+                throw new CharonException("User name is immutable in carbon user store.");
+            }
```

Comment:
Can we move this to a new method? Such as `validateUserExistance` WDYT?

Reply:
addressed in https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/266/commits/d419b5d649f8bc9093b6c4143d0fa188cce5ad99